pkg/apis/workflow/v1alpha1/workflow_types.go

```diff
@@ -1140,11 +1140,55 @@ type WorkflowStep struct {
 	// OnExit is a template reference which is invoked at the end of the
 	// template, irrespective of the success, failure, or error of the
 	// primary template.
+	// DEPRECATED: Use Hooks[exit].Template instead.
 	OnExit string `json:"onExit,omitempty" protobuf:"bytes,11,opt,name=onExit"`
+
+	// Hooks holds the lifecycle hook which is invoked at lifecycle of
+	// step, irrespective of the success, failure, or error status of the primary step
+	Hooks LifecycleHooks `json:"hooks,omitempty" protobuf:"bytes,12,opt,name=hooks"`
+}
+
+type LifecycleEvent string
+
+const (
+	ExitLifecycleEvent = "exit"
+)
+
+type LifecycleHooks map[LifecycleEvent]LifecycleHook
+
+func (lchs LifecycleHooks) GetExitHook() *LifecycleHook {
```

Comment:
I think you might have a bug here. This func will never return nil as currently coded, I think you need this:

```
hook ok := lchs[ExitLifecycleHook]
if ok {
return &hook
} else {
return nil
```